### PR TITLE
fix: Load only databases for which a config exists

### LINKED ISSUE #1450: Bootstrap stuck when rules.pb gets deleted
__Steps to reproduce:__

1. delete a rules.pb file in the object store for which the database directory has other data (e.g. parquet files)
2. restart iox

__Expected behaviour:__

IOx to boot successfully, knowing nothing about the existence of the deleted database.
Possibly warns that there is some data directory without a database config.

__Actual behaviour:__

```
'level=error msg="error getting database config AmazonS3(CloudPath { inner: Parsed(DirsAndFileName { directories: [PathPart(\"1\"), PathPart(\"844910ece80be8bc_infra\")], file_name: Some(PathPart(\"rules.pb\")) }) }) from object store: store error: AWS S3-based Object Store error: Unable to GET data. Bucket: toolsus1.a.influxcloud.net-iox, Location: 1/844910ece80be8bc_infra/rules.pb, Error: The specified key does not exist." target="server" location="server/src/lib.rs:476" time=1620334786114869837
```

repeats every `STORE_ERROR_PAUSE_SECONDS` (100) seconds.

### PR DESCRIPTION
**Pull request by mkmik**
Closes #1450

Longer term, we should use a more sensible object store layout so we don't have to do this kind of contortions.

This hack is needed now because we need to be able to delete S3 files in our staging environment. The current tooling can't spare old rules.pb files, so it's quite possible that we'll hit #1450 soon.

### CODE REVIEW

File: server/src/lib.rs [hunk 1]
```diff
@@ -970,6 +982,24 @@ async fn get_store_bytes(
     Ok(b)
 }
 
+// get the bytes for the database rule config file, if it exists,
+// otherwise it returns none.
+async fn get_database_config_bytes(
+    location: &object_store::path::Path,
+    store: &ObjectStore,
+) -> Result<Option<bytes::BytesMut>> {
```
Comment (alamb) [inline server/src/lib.rs:990]: I wonder if the error handling would be simpler if this always returned an Error

```suggestion
) -> Result<bytes::BytesMut> {
```

And the "error is that no database config is present" was encoded as one of the error types.

this way is fine too, I was just thinking

File: server/src/lib.rs [hunk 2]
```diff
@@ -1122,6 +1177,77 @@ mod tests {
...
+    async fn load_databases() {
+        let temp_dir = TempDir::new().unwrap();
+
+        let store = ObjectStore::new_file(object_store::disk::File::new(temp_dir.path()));
+        let manager = TestConnectionManager::new();
+        let config = config_with_store(store);
+        let server = Server::new(manager, config);
+        server.set_id(ServerId::try_from(1).unwrap()).unwrap();
+
+        create_simple_database(&server, "bananas").await;
```
Comment (alamb) [inline server/src/lib.rs:1213]: 🍌

File: server/src/lib.rs [hunk 3]
```diff
@@ -1122,6 +1177,77 @@ mod tests {
...
+            .expect("cannot delete rules file");
+
+        let manager = TestConnectionManager::new();
+        let config = config_with_store(store);
+        let server = Server::new(manager, config);
+        server.set_id(ServerId::try_from(1).unwrap()).unwrap();
+
+        server.load_database_configs().await.expect("load config");
+
+        assert_eq!(server.db_names_sorted(), vec!["apples"]);
```
Comment (alamb) [inline server/src/lib.rs:1248]: no more bananas. 👍

**Review by alamb — APPROVED** [review]
LGTM

File: server/src/lib.rs (hunk 1 above)
Comment (mkmik) [reply server/src/lib.rs:990]: Yes makes sense. The retry loop would match against that kind of error, warn and return.

I can do it in Monday.

Comment (mkmik) [reply server/src/lib.rs]: done

**Pull request merged**